src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorDiagnosticsPublisher.cs

```diff
@@ -3,268 +3,138 @@
-                ClearClosedDocumentsPublishedDiagnostics(PublishedRazorDiagnostics);
-                ClearClosedDocumentsPublishedDiagnostics(PublishedCSharpDiagnostics);
-
-                _documentClosedTimer?.Dispose();
-                _documentClosedTimer = null;
-
-                if (PublishedRazorDiagnostics.Count > 0 || PublishedCSharpDiagnostics.Count > 0)
-                {
-                    lock (_work)
-                    {
-                        // There's no way for us to know when a document is closed at this layer. Therefore, we need to poll every X seconds
-                        // and check if the currently tracked documents are closed. In practice this work is super minimal.
-                        StartDocumentClosedCheckTimer();
-                    }
-                }
-
-                void ClearClosedDocumentsPublishedDiagnostics<T>(Dictionary<string, IReadOnlyList<T>> publishedDiagnostics) where T : class
-                {
-                    var originalPublishedDiagnostics = new Dictionary<string, IReadOnlyList<T>>(publishedDiagnostics);
-                    foreach (var (key, value) in originalPublishedDiagnostics)
-                    {
-                        Assumes.NotNull(_projectManager);
-                        if (!_projectManager.IsDocumentOpen(key))
-                        {
-                            // Document is now closed, we shouldn't track its diagnostics anymore.
-                            publishedDiagnostics.Remove(key);
-
-                            // If the last published diagnostics for the document were > 0 then we need to clear them out so the user
-                            // doesn't have a ton of closed document errors that they can't get rid of.
-                            if (value.Count > 0)
-                            {
-                                PublishDiagnosticsForFilePath(key, Array.Empty<Diagnostic>());
-                            }
-                        }
-                    }
-                }
-            }
-        }
-        catch
+        foreach (var document in items.GetMostRecentUniqueItems(Comparer.Instance))
         {
-            lock (PublishedRazorDiagnostics)
-            lock (PublishedCSharpDiagnostics)
+            if (token.IsCancellationRequested)
             {
-                _documentClosedTimer?.Dispose();
-                _documentClosedTimer = null;
+                return;
             }
 
-            throw;
+            await PublishDiagnosticsAsync(document, token).ConfigureAwait(false);
         }
     }
 
-    // Internal for testing
-    internal async Task PublishDiagnosticsAsync(IDocumentSnapshot document)
+    private async Task PublishDiagnosticsAsync(IDocumentSnapshot document, CancellationToken token)
     {
         var result = await document.GetGeneratedOutputAsync().ConfigureAwait(false);
+        var csharpDiagnostics = await GetCSharpDiagnosticsAsync(document, token).ConfigureAwait(false);
```

Comment:
Not new, but would it make sense to do these two tasks to get diagnostics in parallel?

Reply:
Only one of the tasks is getting diagnostics. The first task is causing the C# code generation to happen, which I assume is necessary for getting C# diagnostics. The razor diagnostic retrieval below is synchronous.

Reply:
FWIW, I don't know enough about the interaction here know whether they can run in parallel. Because `IDocumentSnapshot.GetGeneratedOutputAsync()` is so (unnecessarily?) complex, it's not clear to me whether there's a subtle interaction that means that `RazorTranslateDiagnositcsService.TranslateAsync(...)` will produce correct results if the generated output hasn't already been computed.

I'm definitely not comfortable making this change until `GetGeneratedOutputAsync()` has been rationalized.

Reply:
I expect `GetGeneratedOutput` here to be a no-op, and the line that uses it (`result.GetCSharpDocument().Diagnostics`) is actually, confusingly, getting the Razor diagnostics. There wouldn't be any harm in doing these things in parallel right now, but without a known cause of perf issues, I wouldn't change it either.